pkg/trace/writer/newwriter/sender.go

```diff
@@ -0,0 +1,458 @@
+	q.enqueueLocked(p)
+
+	if !q.scheduled {
+		// no flush is scheduled; start one
+		q.scheduleFlushLocked(0)
+	}
+}
+
+// scheduleFlushLocked schedules the next flush using the given delay.
+func (q *sender) scheduleFlushLocked(delay time.Duration) {
+	switch {
+	case delay == 0:
+		go q.flush()
+	case q.timer == nil:
+		q.timer = time.AfterFunc(delay, q.flush)
+	default:
+		q.timer.Stop()
+		q.timer.Reset(delay)
+	}
+	q.scheduled = true
+}
+
+// flush drains and sends the entire queue. If anything comes in while flushing
+// or if some of the payloads fail to send as retriable, further follow up flushes
+// are scheduled.
+func (q *sender) flush() {
+	startTime := time.Now()
+
+	// we drain the queue, which is a blocking operation, meaning that
+	// new payloads are stopped from coming in while this happens; but
+	// it is a fast.
+	payloads := q.drainQueue()
+
+	// we send the payloads we've retrieved; while we do this, more payloads
+	// can join the list.
+	done, retries := q.sendPayloads(payloads)
+	if done > 0 {
+		q.recordEvent(eventTypeFlushed, &eventData{
+			count:    int(done),
+			duration: time.Since(startTime),
+		})
+	}
+
+	// we reassess the state of the list and check if further flushes need
+	// to be scheduled
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	switch {
+	case retries > 0:
+		// some sends failed as retriable; we need to back off a bit
+		q.attempt++
+		delay := backoffDuration(q.attempt)
+		q.scheduleFlushLocked(delay)
+	case q.list.Len() > 0:
+		// new items came in while we were flushing; schedule the next flush immediately.
+		q.scheduleFlushLocked(0)
+	default:
+		// everything is flushed
+		q.attempt = 0
+		q.scheduled = false
+	}
+}
+
+// drainQueue drains the entire queue and returns all the payloads that were in it.
+func (q *sender) drainQueue() []*payload {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	var payloads []*payload
+	for q.list.Len() > 0 {
+		v := q.list.Remove(q.list.Front())
+		payloads = append(payloads, v.(*payload))
+	}
+	q.size = 0
+	return payloads
+}
+
+// sendPayloads concurrently sends the given list of payloads. It returns
+// the number of payloads that were added back onto the queue to be retried
+// again later due to an error.
+func (q *sender) sendPayloads(payloads []*payload) (done, retries uint64) {
+	var wg sync.WaitGroup
+	for _, p := range payloads {
+		q.climit <- struct{}{}
+		wg.Add(1)
+		go func(p *payload) {
+			defer func() { <-q.climit }()
+			defer wg.Done()
+			req, err := p.httpRequest(q.cfg.url)
+			if err != nil {
+				log.Errorf("http.Request: %s", err)
+				return
+			}
+			start := time.Now()
+			err = q.do(req)
+			stats := &eventData{
+				bytes:          p.body.Len(),
+				count:          1,
+				duration:       time.Since(start),
+				err:            err,
+				connectionFill: float64(len(q.climit)) / float64(cap(q.climit)),
+			}
+			switch err.(type) {
+			case *retriableError:
+				// request failed again, but can be retried
+				q.enqueueAgain(p)
```

Comment:
Wondering if there's a simpler approach. Could we make this work with only one "pending payloads" channel per sender, and if we need to retry we just throw the payloads back on the same channel? 

If we need to track retries per payload, could wrap it another struct that has a `retries` field that we increment every time it gets retried. 

That way there's only one simple loop reading off of the channel, sending payloads concurrently as they come in (no queueing/flushing), putting them back on the same channel for retries, and adjusting the rate limit according to errors? See my other comment regarding why a rate limit might be a better backoff mechanism than this payload flush strategy.

Reply:
> Wondering if there's a simpler approach. Could we make this work with only one "pending payloads" channel per sender, and if we need to retry we just throw the payloads back on the same channel?

Alex also suggested something like this, but my answer stands as part of my other comment of using a channel as a queue.

> If we need to track retries per payload, could wrap it another struct that has a retries field that we increment every time it gets retried.

An error that can be retried says something about the state of the endpoint (e.g. back off a bit) and is not linked specifically to the payload, so it makes no sense to have a per-payload backoff mechanism. It's either that we are backing off from sending to the endpoint, or we are not. It applies in general to all connections.

> That way there's only one simple loop reading off of the channel, sending payloads concurrently as they come in (no queueing/flushing), putting them back on the same channel for retries, and adjusting the rate limit according to errors? See #3688 (comment) regarding why a rate limit might be a better backoff mechanism than this payload flush-interval strategy.

I will spread my answer to this in that very comment and your others.

Reply:
> An error that can be retried says something about the state of the network (e.g. back off a bit) and is not linked specifically to the payload, so it makes no sense to have a per-payload backoff mechanism.

Good point. Yeah on second thought, I guess no reason to track number of times a payload has been retried due to a retryable error.